src/hdr/hdr_decoder.rs

```diff
@@ -0,0 +1,799 @@
+use Primitive;
+use num_traits::cast::NumCast;
+use num_traits::identities::Zero;
+use std::borrow::Cow;
+use std::error::Error;
+use std::io::{Read, self};
+use std::iter::{Iterator};
+
+use color::{ColorType, Rgb};
+use image::{
+    DecodingResult,
+    ImageDecoder,
+    ImageError,
+    ImageResult,
+};
+
+/// Adapter to conform to ```ImageDecoder``` trait
+#[derive(Debug)]
+pub struct HDRAdapter<R: Read> {
+    inner: Option<HDRDecoder<R>>,
+    meta: HDRMetadata,
+}
+
+impl<R: Read> HDRAdapter<R> {
+    pub fn new(r: R) -> ImageResult<HDRAdapter<R>> {
+        let decoder = try!(HDRDecoder::new(r));
+        let meta = decoder.metadata();
+        Ok(HDRAdapter{ 
+            inner: Some(decoder), 
+            meta: meta,
+        })
+    }
+}
+
+impl<R: Read> ImageDecoder for HDRAdapter<R> {
+    fn dimensions(&mut self) -> ImageResult<(u32, u32)> {
+        Ok((self.meta.width, self.meta.height))
+    }
+
+    fn colortype(&mut self) -> ImageResult<ColorType> {
+        Ok(ColorType::RGB(8))
+    }
+
+    fn row_len(&mut self) -> ImageResult<usize> {
+        Ok(3*(self.meta.width as usize)) // 3 4-byte floats
+    }
+
+    fn read_scanline(&mut self, _: &mut [u8]) -> ImageResult<u32> {
+        unimplemented!()
+    }
+
+    fn read_image(&mut self) -> ImageResult<DecodingResult> {
+        match self.inner.take() {
+            Some(decoder) => {
+                let elem_len = ::std::mem::size_of::<Rgb<u8>>();
+                let mut img: Vec<Rgb<u8>> = try!(decoder.read_image_ldr());
+                // let's transform Vec<Rgb<u8>> into Vec<u8>
+                let p = img.as_mut_ptr() as *mut u8;
+                let len = img.len()*elem_len; // length in bytes
+                let cap = img.capacity()*elem_len; // 
+
+                unsafe {
+                    ::std::mem::forget(img);
+                    Ok(DecodingResult::U8(Vec::from_raw_parts(p, len, cap)))
+                }
+                
+            },
+            None => {
+                Err(ImageError::ImageEnd)
+            }
+        }
+    }
+}
+
+pub const SIGNATURE: &'static [u8] = b"#?RADIANCE";
+const SIGNATURE_LENGTH: usize = 10;
+
+/// An Radiance HDR decoder
+#[derive(Debug)]
+pub struct HDRDecoder<R> {
+    r: R,
+    width: u32,
+    height: u32,
+    meta: HDRMetadata,
+}
+
+/// Refer to [wikipedia](https://en.wikipedia.org/wiki/RGBE_image_format)
+#[repr(C)] #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct RGBE8Pixel {
```

Comment:
Was the requirement for `invert` the only thing preventing you from implementing this as a new `ColorType`?